### PR TITLE
IntentTEDPolicy: Add severity to metadata

### PR DESCRIPTION
**Pull request by dakshvar22**
**Proposed changes**:
- Refactors metadata attached to the prediction of the policy.
- Metadata looks like this now:

```
Metadata schema looks like this:

{
    "query_intent": <metadata of intent that was queried>,
    "ranking": <sorted list of metadata corresponding to all intents.
                Sorting is based on predicted similarities.>
}

Each metadata dictionary looks like this:

{
    "name": <name of intent>,
    "score": <predicted similarity score>,
    "threshold": <threshold used for intent>,
    "severity": <absolute difference between score and threshold>
}
```

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: rasa/core/policies/intent_ted_policy.py [hunk 1]
```diff
@@ -397,26 +403,70 @@ def run_training(
...
+                        Sorting is based on predicted similarities.>
+        }
+
+        Each metadata dictionary looks like this:
+
+        {
+            "name": <name of intent>,
+            "score": <predicted similarity score>,
+            "threshold": <threshold used for intent>,
+            "severity": <absolute difference between score and threshold>
```
Comment (kedz) [inline rasa/core/policies/intent_ted_policy.py:424]: Does severity need to be here if we are additionally storing the numbers to compute it? Not sure what our policy is on the tradeoff cost for storing this info vs recomputing later.

File: rasa/core/policies/intent_ted_policy.py [hunk 2]
```diff
@@ -397,26 +403,70 @@ def run_training(
...
-        Similarities for all intents and their thresholds are attached as metadata.
+        Metadata schema looks like this:
+
+        {
+            "query_intent": <metadata of intent that was queried>,
+            "ranking": <sorted list of metadata corresponding to all intents.
+                        Sorting is based on predicted similarities.>
+        }
+
+        Each metadata dictionary looks like this:
```
Comment (kedz) [inline rasa/core/policies/intent_ted_policy.py:418]: "Each metadata dictionary in the ranking field looks like this:"

At first it wasn't clear to me how the top schema related to the bottom.

Comment (kedz) [reply rasa/core/policies/intent_ted_policy.py:418]: Oh actually I see that both fields contain metadata, where query_intent is a singleton and ranking is a list. Perhaps maybe worth noting that the query_intent metadata is also in ranking?

**Review by kedz — COMMENTED** [review]
I think in the discussion when we talked about severity, I was assuming it was only being calculated for intents which triggered `action_unlikely_intent`, but it looks like its getting computed for all intents. As its now calculated, If one intent is above the threshold but equally distant to it as an intent below the threshold, wont they be assigned the same severity? This seems confusing to me if so.

Why not just calculate severity as threshold - score? This would give intents above the threshold negative severity. Or alternatively max(0, threshold - score) and intents above the threshold have 0 severity?

**Review by JEM-Mosig — APPROVED** [review]
Looks good!

**Comment by JEM-Mosig** [comment]
> Why not just calculate severity as threshold - score? This would give intents above the threshold negative severity. Or alternatively max(0, threshold - score) and intents above the threshold have 0 severity?

Good point! I think `threshold - score` is better.

File: rasa/core/policies/intent_ted_policy.py (hunk 1 above)
Comment (dakshvar22) [reply rasa/core/policies/intent_ted_policy.py:424]: Computation of severity is a property of `IntentTEDPolicy`, so it should definitely stay inside it and added to the metadata.
On the tradeoff: Users can always use `ranking_length` parameter to reduce the ranking list if they see that exploding the memory.

Comment (dakshvar22) [reply rasa/core/policies/intent_ted_policy.py:424]: Thanks for bringing that up, support for `ranking_length` was missing and I added it now.

**Comment by dakshvar22** [comment]
Agree about changing the severity measure. Changed it to `threshold - score`

File: rasa/core/policies/intent_ted_policy.py (hunk 2 above)
Comment (dakshvar22) [reply rasa/core/policies/intent_ted_policy.py:421]: > Perhaps maybe worth noting that the query_intent metadata is also in ranking?

Added that in.

**Review by kedz — APPROVED** [review]
Looks good!

**Pull request merged**